src/context/GlobalContext.tsx

```diff
@@ -0,0 +1,62 @@
+import React, {
+  useContext,
+  useReducer,
+  useEffect,
+  createContext,
+  useMemo,
+} from "react";
+import {
+  ContextActions,
+  GlobalState,
+  Types,
+} from "src/context/context.interface";
+import * as ActionCreators from "src/context/ActionCreators";
+import { rootReducer } from "src/context/reducer";
+import { AuthFormData } from "src/common/interface";
+import { auth } from "src/modules/AuthService";
+import { User } from "src/modules/User";
+
+const initialState: GlobalState = {
+  loading: true,
+};
+const GlobalContext = createContext<{
+  state: GlobalState;
+  actions?: ContextActions;
+}>({
+  state: initialState,
+});
+
+export function useAuth() {
+  return useContext(GlobalContext);
+}
+
+export const GlobalProvider: React.FC = ({ children }) => {
+  const [state, dispatch] = useReducer(rootReducer, initialState);
+
+  const actions = useMemo<ContextActions>(
+    () => ({
+      signUp: (payload: AuthFormData) =>
+        ActionCreators.register(dispatch, payload),
+      signIn: (payload: AuthFormData) =>
+        ActionCreators.login(dispatch, payload),
+      signOut: () => ActionCreators.logout(dispatch),
+    }),
+    []
+  );
+
+  useEffect(() => {
+    const subscribeAuthChange = auth.onAuthStateChanged((user) => {
+      dispatch({
+        type: Types.SET_USER,
+        payload: user ? new User(user) : undefined,
+      });
+    });
+    return subscribeAuthChange;
```

Comment:
Is this the `unsubscribe` handler?

Reply:
yes

Reply:
let's rename the variable to reflect that

Reply:
Yes, it is for handling when component unmounts to remove this listener.

Reply:
I think it is clear that we return cleanup function  and `subscribeAuthChange` is pretty straight forward function name for for firebase auth listener

Reply:
Let's go with `unsubscribeAuthChange`